File: chrome/browser/previews/android/previews_android_bridge.cc
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/previews/android/previews_android_bridge.h"

#include <memory>

#include "base/android/jni_android.h"
#include "chrome/browser/previews/previews_ui_tab_helper.h"
#include "content/public/browser/web_contents.h"
#include "jni/PreviewsAndroidBridge_jni.h"

static jlong JNI_PreviewsAndroidBridge_Init(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  return reinterpret_cast<intptr_t>(new PreviewsAndroidBridge(env, obj));
}

PreviewsAndroidBridge::PreviewsAndroidBridge(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj)
    : weak_factory_(this) {}

PreviewsAndroidBridge::~PreviewsAndroidBridge() {}

jboolean PreviewsAndroidBridge::ShouldShowPreviewUI(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj,
    const base::android::JavaParamRef<jobject>& j_web_contents) {
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(j_web_contents);
  if (!web_contents)
    return false;

  PreviewsUITabHelper* tab_helper =
      PreviewsUITabHelper::FromWebContents(web_contents);
  if (!tab_helper)
    return false;

  return tab_helper->should_display_android_omnibox_badge();
}

File: chrome/android/java/src/org/chromium/chrome/browser/previews/PreviewsAndroidBridge.java
// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.previews;

import org.chromium.content_public.browser.WebContents;

/**
 * Java bridge class to C++ Previews code.
 */
public final class PreviewsAndroidBridge {
    private static PreviewsAndroidBridge sBridge;

    public static PreviewsAndroidBridge getInstance() {
        if (sBridge == null) {
            sBridge = new PreviewsAndroidBridge();
        }
        return sBridge;
    }

    private final long mNativePreviewsAndroidBridge;

    private PreviewsAndroidBridge() {
        mNativePreviewsAndroidBridge = nativeInit();
    }

    public boolean shouldShowPreviewUI(WebContents webContents) {
        return nativeShouldShowPreviewUI(mNativePreviewsAndroidBridge, webContents);
    }

    public String getOriginalHost(WebContents webContents) {
        assert shouldShowPreviewUI(webContents) : "getOriginalHost called on a non-preview page";
        return nativeGetOriginalHost(mNativePreviewsAndroidBridge, webContents);
    }

    public void loadOriginal(WebContents webContents) {
        assert shouldShowPreviewUI(webContents) : "loadOriginal called on a non-preview page";
        nativeLoadOriginal(mNativePreviewsAndroidBridge, webContents);
    }

    private native long nativeInit();
    private native boolean nativeShouldShowPreviewUI(
            long nativePreviewsAndroidBridge, WebContents webContents);
    private native String nativeGetOriginalHost(
            long nativePreviewsAndroidBridge, WebContents webContents);
    private native void nativeLoadOriginal(
            long nativePreviewsAndroidBridge, WebContents webContents);
}
